lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,239 @@
+                  fragment.insert(0, html_comment)
+                  break
+                end
+
+                body.close if body.respond_to?(:close)
+
+                update_content_length(headers, html_comment)
+                # ensure idempotency on injection in case middleware is inserted or called twice
+                env[RUM_INJECTION_FLAG] = true
+
+                updated_response = ::Rack::Response.new(body, status, headers)
+                Datadog.logger.debug('Rum injection successful')
+                return updated_response.finish
+              else
+                Datadog.logger.debug('Rum injection unsuccessful')
+                return result
+              end
+            end
+
+            # catchall if an earlier conditional is not met
+            result
+          rescue StandardError => e
+            Datadog.logger.warn("error checking rum injectability #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+            # we should ensure we don't interfere if original app response if our injection code has an exception
+            result
+          end
+        end
+
+        private
+
+        def configuration
+          Datadog.configuration[:rack]
+        end
+
+        def should_inject?(headers, env, status)
+          status == 200 &&
+            !env[RUM_INJECTION_FLAG] &&
+            no_cache?(headers, env) &&
+            !compressed?(headers) &&
+            !attachment?(headers) &&
+            !streaming?(headers, env) &&
+            injectable_html?(headers)
+        # catch everything and swallow it here for defensiveness
+        rescue Exception => e # rubocop:disable Lint/RescueException
+          Datadog.logger.warn("Error during rum injection  #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+          return nil
+        end
+
+        def compressed?(headers)
+          headers && headers.key?(CONTENT_ENCODING_HEADER) &&
+            !headers[CONTENT_ENCODING_HEADER].start_with?(IDENTITY)
+        end
+
+        def injectable_html?(headers)
+          (headers && headers.key?(CONTENT_TYPE_HEADER) && !headers[CONTENT_TYPE_HEADER].nil?) &&
+            (headers[CONTENT_TYPE_HEADER].include?(HTML_CONTENT) ||
+            headers[CONTENT_TYPE_HEADER].include?(XHTML_CONTENT))
+        end
+
+        def attachment?(headers)
+          (headers && headers.key?(CONTENT_DISPOSITION_HEADER) && !headers[CONTENT_DISPOSITION_HEADER].nil?) &&
+            !headers[CONTENT_DISPOSITION_HEADER].include?(INLINE)
+        end
+
+        def streaming?(headers, env)
+          # https://api.rubyonrails.org/classes/ActionController/Streaming.html
+          # rails recommends disabling middlewares that interact with response body
+          # when streaming via ActionController::Streaming
+          # TODO: if required to patch streaming, investigate patching further upstream, in the render action perhaps
+          return true if (headers && headers.key?(TRANSFER_ENCODING_HEADER) &&
```

Comment:
I see we are checking for the presence of `headers` many times (`if headers && ...`).
I think we should have a higher-level short circuit that skips pretty much all of this class if `headers` is not available.

Reply:
makes sense, tried to clean this up